### ISSUE TITLE
Incorrect type inference for list comprehension returning List[List[Union]]

### ISSUE THREAD

**Issue by wplohrmann** [python/mypy#13439]
The following snippet illustrates the problem:

```python
from typing import List, Union

def wow() -> List[str]:
    return ["wow"]

def wow_flat() -> str:
    return "wow"

insert_str = [True, False]

doesnt_work: List[List[Union[str, int]]] = [
    wow() if x else [0] for x in insert_str
]

works_1: List[List[Union[str, int]]] = [
    ["wow"] if x else [0] for x in insert_str
]

works_2: List[Union[str, int]] = [
    wow_flat() if x else 0 for x in insert_str
]

works_3: List[Union[str, int]] = [
    "wow" if x else 0 for x in insert_str
]
```

I expected it to produce no type errors but for some reason it throws an error on line 12:
```
example.py:12: error: List comprehension has incompatible type List[List[str]]; expected List[List[Union[str, int]]]
example.py:12: error: List item 0 has incompatible type "int"; expected "str"
```

The issue seems to be related to both the doubly nested list as well as the fact that we're using a function to produce a `List[str]` object. If you remove either of those elements everything works fine.

##  Environment
- Mypy version used: 0.960 (compiled: yes)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.9.13
- Operating system and version: Mac 12.4

**Comment by wplohrmann** [comment]
The issue persists if I change the type of `doesnt_work` to `List[Union[List[str], List[int]]]` but then curiously I only get the following error:

```
example.py:12: error: List item 0 has incompatible type "int"; expected "str"
```

**Comment by wplohrmann** [comment]
The following works:

```python
works_now: List[Union[List[str], List[int]]] = []
for x in insert_str:
    if x:
        works_now.append(wow())
    else:
        works_now.append([0])
```

**Comment by erictraut** [comment]
Mypy's behavior is correct here. The `doesnt_work` case is a type violation because of the invariance of `list`. The `wow` function returns a value of type `list[str]`, and a `list[str]` is not allowed as an element in a `list[list[str | int]]`.

I recommend closing the bug unless there's a desire to improve the error message, in which case the title should probably be updated.
